R/glossary.R

```diff
@@ -0,0 +1,398 @@
+    abort("Every `glossary` entry must have type character and length 1.")
+  }
+  glossary <- Map(x, names(x), f = function(def, label) {
+    data <- c(definition = def, abbreviation = label)
+    as_glossary_entry(data)
+  })
+  class(glossary) <- c("tex_glossary", "glossary", "list")
+  glossary
+}
+
+as_glossary_entry <- function(x) {
+  x <- as.list(x)
+  class(x) <- c("glossary_entry", "list")
+  x
+}
+
+abort_bad_glo_labels <- function(x, what) {
+  if(!all(what %in% names(x))) {
+    bad <- setdiff(what, names(x))
+    names(bad) <- rep("x", length(bad))
+    msg <- c("Requested definitions not found in glossary object", bad)
+    abort(msg)
+  }
+}
+
+#' Return formatted notes from a tex glossary file or glossary object
+#'
+#' @param x path to a tex glossary file, a glossary object, or a list that can
+#' be coerced to a glossary object with [as_glossary()].
+#'
+#' @inheritParams read_glossary
+#' @inheritParams st_notes_glo
+#'
+#' @examples
+#' file <- system.file("glo", "glossary.tex", package = "pmtables")
+#'
+#' glossary_notes(file, WT, CRCL)
+#'
+#' g <- as_glossary(ss = "steady state", ALB = "albumin", WT = "weight")
+#'
+#' glossary_notes(g, ALB, ss)
+#'
+#' what <- c("WT", "NCA", "NPDE")
+#'
+#' g <- read_glossary(file)
+#'
+#' glossary_notes(g, tidyselect::all_of(what))
+#'
+#' l <- list(ss = "steady state", ALB = "albumin", WT = "weight")
+#'
+#' glossary_notes(l, ALB, ss)
+#'
+#' @seealso [st_notes_glo()], [read_glossary()]
+#' @export
+glossary_notes <- function(x, ...) UseMethod("glossary_notes")
+
+#' @param ... unquoted glossary labels to be included in the notes text.
+#' @param format passed to [read_glossary()].
+#' @rdname glossary_notes
+#' @export
+glossary_notes.character <- function(x, ..., format = guess_glo_fmt(x)) {
+  x <- read_glossary(x, format = format)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.list <- function(x, ...) {
+  x <- as_glossary(x)
+  glossary_notes(x, ...)
+}
+
+#' @rdname glossary_notes
+#' @export
+glossary_notes.glossary <- function(x, ..., sep = ": ", collapse = "; ") {
+  labels <- names(eval_select(expr(c(...)), x))
+  if(!length(labels)) labels <- names(x)
+  build_glossary_notes(x, labels, sep, collapse)
+}
+
+build_glossary_notes <- function(glossary, labels, sep, collapse) {
+  abort_bad_glo_labels(glossary, labels)
+  glossary <- glossary[labels]
+  abb <- vapply(glossary, "[[", "abbreviation", FUN.VALUE = "a")
+  notes <- lapply(glossary, "[[", "definition")
+  notes <- paste0(abb, sep, notes)
+  if(is.character(collapse)) {
+    notes <- paste0(notes, collapse = collapse)
+  }
+  notes
+}
+
+#' @export
+print.glossary_entry <- function(x, ...) {
+  cat(paste0(x$definition, " (", x$abbreviation, ")\n"))
+}
+
+#' @export
+print.glossary <- function(x, ...) {
+  if(!length(x)) {
+    cat("No glossary entries found.")
+    return(invisible(NULL))
+  }
+  label <- names(x)
+  def <- map_chr(x, "definition")
```

Comment:
The case I mentioned [here](https://github.com/metrumresearchgroup/pmtables/pull/326#discussion_r1621588141) will still fail:

```R
as_glossary(ss = "steady state")[2]
#> Error in `map_chr()` at pmtables/R/glossary.R:311:3:
#> ℹ In index: 1.
#> Caused by error:
#> ! Result must be length 1, not 0.
#> Run `rlang::last_trace()` to see where the error occurred
```

The length guard won't catch that because of this named-list behavior:

```R
x <- list(foo = "foo")
x[2]
#> $<NA>
#> NULL

length(x[2])
#> [1] 1
```

How about this guard instead?

```diff
diff --git a/R/glossary.R b/R/glossary.R
index 76fc618..3016dc9 100644
--- a/R/glossary.R
+++ b/R/glossary.R
@@ -303,8 +303,8 @@ print.glossary_entry <- function(x, ...) {

 #' @export
 print.glossary <- function(x, ...) {
-  if(!length(x)) {
-    cat("No glossary entries found.")
+  if(is.null(unlist(x))) {
+    cat("No glossary entries found.\n")
     return(invisible(NULL))
   }
   label <- names(x)
```